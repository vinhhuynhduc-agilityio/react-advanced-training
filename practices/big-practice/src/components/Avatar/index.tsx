import React, { memo } from 'react';
import clsx from 'clsx';

interface AvatarProps {
  src: string;
  alt?: string;
  size?: string;
  ariaLabel?: string;
}

export const Avatar: React.FC<AvatarProps> = memo(({
  src,
  alt = 'Avatar',
  size = 'w-10 h-10',
  ariaLabel = '',
}) => {
  const validSizes = ['w-10 h-10', 'w-8 h-8', 'w-24 h-24', 'w-14 h-14'];

  return (
    <img
      src={src}
      alt={alt}
      className={clsx(
        'rounded-full object-cover',
        validSizes.includes(size) ? size : 'w-10 h-10'
      )}
      aria-label={ariaLabel}
    />
  );
});
