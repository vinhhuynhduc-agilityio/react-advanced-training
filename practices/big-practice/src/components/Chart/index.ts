export * from './ChartIndividualEmployeeProgress';
export * from './ChartTotalTasksByProjects';
export * from './ChartTotalTasksCompleted';