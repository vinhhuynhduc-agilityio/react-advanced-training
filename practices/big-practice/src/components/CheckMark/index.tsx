import React, { useState, useEffect } from "react";
import { FaCheckCircle, FaClock } from "react-icons/fa";

export interface IconRendererProps {
  onStatusValueChange: () => void;
  value: boolean;
  isDisabled?: boolean;
}

export const CheckMark: React.FC<IconRendererProps> = (props) => {
  const [isComplete, setIsComplete] = useState<boolean>(props.value);
  const { isDisabled, onStatusValueChange } = props;

  // Update icon state when props.value changes
  useEffect(() => {
    setIsComplete(props.value);
  }, [props.value]);

  // Predefined variables for clarity
  const isDisabledClass = isDisabled
    ? "opacity-50 pointer-events-none"
    : "cursor-pointer";

  return (
    <span
      data-testid="status-icon"
      className={`flex justify-center items-center w-full h-full ${isDisabledClass}`}
      onClick={onStatusValueChange}
      aria-label="icon"
    >
      {isComplete ? (
        <FaCheckCircle data-testid="icon-check" color="#1CA1C1" size={20} />
      ) : (
        <FaClock data-testid="icon-clock" color="#FF5C4C" size={20} />
      )}
    </span>
  );
};
