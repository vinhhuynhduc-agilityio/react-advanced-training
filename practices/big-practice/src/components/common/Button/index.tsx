import React from 'react';

interface ButtonProps {
  label: string;
  onClick?: () => void;
  type?: 'button' | 'submit' | 'reset';
  variant?: 'primary' | 'secondary' | 'default';
  disabled?: boolean;
  ariaLabel?: string;
}

const buttonClasses = {
  default: 'px-4 py-2 border-none mr-7 bg-[#F4F5F9] text-[#1CA1C1] hover:bg-slate-300',
  primary: 'bg-slate-200 text-blue-600 font-bold',
  secondary: 'bg-slate-200 text-pink-600 font-bold',
  disabled: 'px-4 py-2 border-none mr-7 bg-gray-300 text-gray-500 cursor-not-allowed',
};

export const Button: React.FC<ButtonProps> = ({
  label,
  onClick,
  type = 'button',
  variant = 'default',
  disabled = false,
  ariaLabel,
}) => {

  return (
    <button
      type={type}
      onClick={onClick}
      disabled={disabled}
      className={buttonClasses[variant]}
      aria-label={ariaLabel}
    >
      {label}
    </button>
  );
};
