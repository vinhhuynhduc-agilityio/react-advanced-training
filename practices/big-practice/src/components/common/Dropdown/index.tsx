import {
  useState,
  useRef,
  useEffect,
  forwardRef
} from 'react';
import clsx from 'clsx';

interface OptionData {
  id: string;
  value: string;
};

interface DropdownProps {
  options: OptionData[];
  placeholder: string;
  onSelect: (option: OptionData) => void;
};

export const Dropdown = forwardRef<HTMLInputElement, DropdownProps>(
  ({
    options,
    placeholder,
    onSelect
  },
    ref
  ) => {
    const [isOpen, setIsOpen] = useState(false);
    const [selectedOption, setSelectedOption] = useState<OptionData | null>(null);

    const dropdownRef = useRef<HTMLDivElement | null>(null);

    const handleSelect = (option: OptionData) => {
      setSelectedOption(option);
      onSelect(option);
      setIsOpen(false);
    };

    const toggleDropdown = () => {
      setIsOpen(!isOpen);
    };

    const handleDropdownClose = (event: Event) => {
      if (event.type === 'resize') {
        setIsOpen(false);
        return;
      }

      if (
        dropdownRef.current &&
        !dropdownRef.current.contains(event.target as Node)
      ) {
        setIsOpen(false);
      }
    };

    useEffect(() => {
      // Event listener for mouse click outside
      document.addEventListener('mousedown', handleDropdownClose);

      // Event listener for window resize
      window.addEventListener('resize', handleDropdownClose);

      // Cleanup
      return () => {
        document.removeEventListener('mousedown', handleDropdownClose);
        window.removeEventListener('resize', handleDropdownClose);
      };
    }, []);

    const valueDropdown = selectedOption
      ? selectedOption.value.toString()
      : '';

    return (
      <div
        className="relative w-full"
        ref={dropdownRef}
      >
        <div
          className="flex items-center border border-gray-300 rounded-md p-2 cursor-text bg-white"
          onClick={toggleDropdown}
        >
          <input
            ref={ref}
            type="text"
            placeholder={placeholder}
            className="flex-1 outline-none bg-transparent text-gray-700 cursor-text"
            readOnly
            value={valueDropdown}
          />
          <span className="ml-2 w-0 h-0 border-l-4 border-r-4 border-t-4 border-t-gray-500 border-l-transparent border-r-transparent cursor-pointer"></span>
        </div>
        {isOpen && (
          <div className="absolute top-full left-0 right-0 max-h-[297px] bg-white border border-gray-300 rounded-md shadow-lg overflow-y-auto z-10">
            {options.map((option) => (
              <div
                key={option.id}
                role="option"
                className={clsx(
                  "p-1.5 cursor-pointer text-gray-700 hover:bg-gray-100 hover:text-blue-500 border-b border-gray-300 last:border-0",
                  selectedOption?.id === option.id && "bg-blue-100"
                )}
                onClick={() => handleSelect(option)}
              >
                {option.value}
              </div>
            ))}
          </div>
        )}
      </div>
    );
  }
);
