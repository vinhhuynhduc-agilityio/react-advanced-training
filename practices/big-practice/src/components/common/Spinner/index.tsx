interface SpinnerProps {
  borderColor?: string;
}

export const Spinner: React.FC<SpinnerProps> = ({
  borderColor = "border-t-blue-500",
}) => {
  const validBorderColors = [
    "border-t-blue-500",
    "border-t-red-500",
    "border-t-green-500",
    "border-t-yellow-500",
  ];

  const spinnerBorderClass = validBorderColors.includes(borderColor)
    ? borderColor
    : "border-t-blue-500";

  return (
    <div className="flex items-center justify-center w-full h-full">
      <div
        className={`w-12 h-12 border-4 border-gray-300 ${spinnerBorderClass} rounded-full animate-spin`}
      ></div>
    </div>
  );
};
