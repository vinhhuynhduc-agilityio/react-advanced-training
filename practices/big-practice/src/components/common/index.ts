export * from './Button';
export * from './Dropdown';
export * from './Spinner';
