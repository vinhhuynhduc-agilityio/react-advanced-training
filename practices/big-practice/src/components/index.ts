export * from './ChartIndividualEmployeeProgress';
export * from './ChartTotalTasksByProjects';
export * from './ChartTotalTasksCompleted';
export * from './ModalDialog';
export * from './ProjectForm';
export * from './TaskForm';
export * from './UserProfileForm';
export * from './Header';
export * from './Footer';
export * from './DataGrid';

export { default as TaskDashboard } from './TaskDashboard';
export { default as UserListDrawer } from './UserListDrawer';
