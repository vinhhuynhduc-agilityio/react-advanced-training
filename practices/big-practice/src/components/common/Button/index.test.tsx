import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Button } from '@/components/common';

describe('Button component', () => {
  it('matches snapshot for default state', () => {
    const { container } = render(
      <Button
        label="Click me"
        onClick={() => { }} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders the correct label', () => {
    render(
      <Button
        label="Submit"
        onClick={() => { }} />
    );
    expect(screen.getByText('Submit')).toBeInTheDocument();
  });

  it('renders with the correct type', () => {
    const { container } = render(
      <Button
        label="Reset"
        type="reset"
        onClick={() => { }} />
    );
    const button = container.querySelector('button');
    expect(button).toHaveAttribute('type', 'reset');
  });

  it('calls the onClick handler when clicked', () => {
    const onClickMock = jest.fn();
    render(
      <Button
        label="Click me"
        onClick={onClickMock} />
    );

    const button = screen.getByText('Click me');
    fireEvent.click(button);
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it('applies correct class for default variant', () => {
    const { container } = render(
      <Button
        label="Default"
        onClick={() => { }} />
    );
    const button = container.querySelector('button');
    expect(button).toHaveClass('px-4 py-2 border-none mr-7 bg-[#F4F5F9] text-[#1CA1C1] hover:bg-slate-300');
  });

  it('applies correct class for primary variant', () => {
    const { container } = render(
      <Button
        label="Primary"
        variant="primary"
        onClick={() => { }} />
    );
    const button = container.querySelector('button');
    expect(button).toHaveClass('bg-slate-200 text-blue-600 font-bold');
  });

  it('applies correct class for secondary variant', () => {
    const { container } = render(
      <Button
        label="Secondary"
        variant="secondary"
        onClick={() => { }} />
    );
    const button = container.querySelector('button');
    expect(button).toHaveClass('bg-slate-200 text-pink-600 font-bold');
  });

  it('does not call onClick when button is disabled', () => {
    const onClickMock = jest.fn();
    render(
      <Button
        label="Disabled"
        onClick={onClickMock}
        disabled />
    );

    const button = screen.getByText('Disabled');
    fireEvent.click(button);
    expect(onClickMock).toHaveBeenCalledTimes(0);
  });

  it('applies aria-label if provided', () => {
    render(
      <Button
        label="Accessible Button"
        ariaLabel="Submit Form"
        onClick={() => { }} />
    );
    const button = screen.getByLabelText('Submit Form');
    expect(button).toBeInTheDocument();
  });
});
