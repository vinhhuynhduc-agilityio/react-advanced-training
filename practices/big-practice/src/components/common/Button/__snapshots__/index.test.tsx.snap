// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot for default state 1`] = `
<div>
  <button
    class="px-4 py-2 border-none mr-7 bg-[#F4F5F9] text-[#1CA1C1] hover:bg-slate-300"
    type="button"
  >
    Click me
  </button>
</div>
`;
